contracts/lib/OnChainIdentity.sol

```diff
@@ -0,0 +1,228 @@
+        // This empty reserved space is put in place to allow future versions
+        // of the SMT library to add new Data struct fields without shifting down
+        // storage of upgradable contracts that use this struct as a state variable
+        // (see https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable#storage-gaps)
+        uint256[45] __gap;
+    }
+
+    /**
+     * @dev SMT addresses
+     */
+    struct TreeRoots {
+        SmtLib.Data claimsRoot;
+        SmtLib.Data revocationsRoot;
+        SmtLib.Data rootsRoot;
+    }
+
+    /**
+     * @dev roots used in last State Transition
+     */
+    struct LastTreeRoots {
+        uint256 lastClaimsTreeRoot;
+        uint256 lastRevocationsTreeRoot;
+        uint256 lastRootsTreeRoot;
+    }
+
+     function initialize(
+        IdentityData storage self,
+        address _stateContractAddr,
+        uint256 maxDepth,
+        TreeRoots storage treeRoots
+    ) external {
+        
+        self.stateContract = IState(_stateContractAddr);
+        self.isOldStateGenesis = true;
+
+        treeRoots.claimsRoot.initialize(maxDepth);
+        treeRoots.revocationsRoot.initialize(maxDepth);
+        treeRoots.rootsRoot.initialize(maxDepth);
+
+        self.identityId = GenesisUtils.calcOnchainIdFromAddress(0x0212, address(this));
+    }
+
+    /**
+     * @dev Add claim
+     * @param claim - claim data
+     */
+    function addClaim(TreeRoots storage self, uint256[8] memory claim) external {
+        uint256[4] memory claimIndex;
+        uint256[4] memory claimValue;
+        for (uint8 i = 0; i < 4; i++) {
+            claimIndex[i] = claim[i];
+            claimValue[i] = claim[i + 4];
+        }
+        uint256 hashIndex = PoseidonUnit4L.poseidon(claimIndex);
+        uint256 hashValue = PoseidonUnit4L.poseidon(claimValue);
+        self.claimsRoot.addLeaf(hashIndex, hashValue);
+    }
+
+     /**
+     * @dev Add claim hash
+     * @param hashIndex - hash of claim index part
+     * @param hashValue - hash of claim value part
+     */
+    function addClaimHash(TreeRoots storage self, uint256 hashIndex, uint256 hashValue) external {
+        self.claimsRoot.addLeaf(hashIndex, hashValue);
+    }
+
+     /**
+     * @dev Revoke claim using it's revocationNonce
+     * @param revocationNonce - revocation nonce
+     */
+    function revokeClaim(TreeRoots storage self, uint64 revocationNonce) external {
+        self.revocationsRoot.addLeaf(uint256(revocationNonce), 0);
+    }
+
+    /**
+     * @dev Make state transition
+     */
+    function transitState(TreeRoots storage self, LastTreeRoots storage lastTreeRoots, IdentityData storage identity) external {
+        uint256 currentClaimsTreeRoot = self.claimsRoot.getRoot();
+        uint256 currentRevocationsTreeRoot = self.revocationsRoot.getRoot();
+        uint256 currentRootsTreeRoot = self.rootsRoot.getRoot();
+
+        require(
+            (lastTreeRoots.lastClaimsTreeRoot != currentClaimsTreeRoot) ||
```

Comment:
`lastTreeRoots.lastClaimsTreeRoot`. Can we refactoring to `lastTreeRoots.ClaimsRoot`. Becasue a lot of duplication words `last`, `tree` and `roots`.